src/dune_rules/cram/cram_exec.ml

```diff
@@ -436,24 +456,38 @@ let run ~env ~script lexbuf : string Fiber.t =
   let sanitized = sanitize ~parent_script:sh_script.script raw in
   compose_cram_output sanitized
 
-let run ~env ~script =
-  run_expect_test script ~f:(fun lexbuf -> run ~env ~script lexbuf)
+let run ~env ~script ~shell_spec =
+  run_expect_test script ~f:(fun lexbuf -> run ~env ~script ~shell_spec lexbuf)
+
+type ('path, _) spec =
+  { script : 'path
+  ; shell_spec : 'path Shell_spec.t
+  }
 
 module Spec = struct
-  type ('path, _) t = 'path
+  type ('path, 'target) t = ('path, 'target) spec
 
   let name = "cram"
 
   let version = 1
 
-  let bimap path f _ = f path
+  let bimap spec f _ =
+    { script = f spec.script; shell_spec = Shell_spec.conv f spec.shell_spec }
 
   let is_useful_to ~distribute:_ ~memoize:_ = true
 
-  let encode script path _ : Dune_lang.t =
-    List [ Dune_lang.atom_or_quoted_string "cram"; path script ]
+  let encode { script; shell_spec } path _ : Dune_sexp.t =
+    let atom = Dune_lang.atom_or_quoted_string in
+    let sh : Dune_lang.t =
+      match shell_spec with
```

Comment:
Actually rather than having 3 shells here. Why not immediately `which` them? You should also validate the custom shell as actually existing. Something else that comes to mind is that users may want to build their own shells, but then `Bin.which` will not find them. We should probably document this limitation somewhere, however I am not sure if it is even worth addressing since it is such a niche concern.

Reply:
Good point. I'll see how to implement it.

Reply:
addressed in 071f56c82